apps/src/code-studio/components/libraries/LibraryPublisher.jsx

```diff
@@ -0,0 +1,224 @@
+/*global dashboard*/
+import React from 'react';
+import PropTypes from 'prop-types';
+import libraryParser from './libraryParser';
+import i18n from '@cdo/locale';
+import color from '@cdo/apps/util/color';
+import {Heading2} from '@cdo/apps/lib/ui/Headings';
+import Button from '@cdo/apps/templates/Button';
+
+const styles = {
+  alert: {
+    color: color.red,
+    width: '90%',
+    paddingTop: 8
+  },
+  largerCheckbox: {
+    width: 20,
+    height: 20,
+    marginLeft: 0,
+    marginRight: 10,
+    marginTop: 10,
+    marginBottom: 10
+  },
+  info: {
+    fontSize: 12,
+    fontStyle: 'italic'
+  },
+  textInput: {
+    fontSize: 14,
+    padding: 6,
+    color: color.dimgray
+  },
+  description: {
+    width: '98%',
+    resize: 'vertical'
+  }
+};
+
+/**
+ * @readonly
+ * @enum {string}
+ */
+export const PublishState = {
+  DEFAULT: 'default',
+  ERROR_PUBLISH: 'error_publish',
+  INVALID_INPUT: 'invalid_input'
+};
+
+/**
+ * An interactive page for a dialog that can be used to publish or unpublish
+ * a library from a source project.
+ */
+export default class LibraryPublisher extends React.Component {
+  static propTypes = {
+    onPublishSuccess: PropTypes.func.isRequired,
+    libraryDetails: PropTypes.object.isRequired,
+    libraryClientApi: PropTypes.object.isRequired
+  };
+
+  state = {
+    publishState: PublishState.DEFAULT,
+    libraryName: libraryParser.suggestName(
+      this.props.libraryDetails.libraryName
+    ),
+    libraryDescription: '',
+    selectedFunctions: {}
+  };
+
+  setLibraryName = event => {
+    let sanitizedName = libraryParser.sanitizeName(event.target.value);
+    if (sanitizedName === this.state.libraryName) {
+      return;
+    }
+    this.setState({libraryName: sanitizedName});
+  };
+
+  publish = () => {
+    const {libraryDescription, libraryName, selectedFunctions} = this.state;
+    const {librarySource, sourceFunctionList} = this.props.libraryDetails;
+    const {libraryClientApi, onPublishSuccess} = this.props;
+    let functionsToPublish = sourceFunctionList.filter(sourceFunction => {
+      return selectedFunctions[sourceFunction.functionName];
+    });
+
+    if (!(libraryDescription && functionsToPublish.length > 0)) {
+      this.setState({publishState: PublishState.INVALID_INPUT});
+      return;
+    }
+
+    let libraryJson = libraryParser.createLibraryJson(
+      librarySource,
+      functionsToPublish,
+      libraryName,
+      libraryDescription
+    );
+
+    // TODO: Display final version of error and success messages to the user.
```

Comment:
Is there anything left to do related to this TODO?

Reply:
✔️